File: problems/2025/09-16/postfix-evaluation/solution.js
function evaluatePostfix(arr) {
  const st = [];
  for (const tok of arr) {
    if (tok === "+" || tok === "-" || tok === "*" || tok === "/" || tok === "^") {
      const b = st.pop();
      const a = st.pop();
      switch (tok) {
        case "+": st.push(a + b); break;
        case "-": st.push(a - b); break;
        case "*": st.push(a * b); break;
        case "/": {
          // floor division toward -infinity
          const q = Math.floor(a / b);
          st.push(q);
          break;
        }
        case "^": st.push(Math.pow(a, b) | 0); break;
      }
    } else {
      st.push(parseInt(tok, 10));
    }
  }
  return st[st.length - 1];
}

console.log(evaluatePostfix(["2","3","1","*","+","9","-"])); // -4
console.log(evaluatePostfix(["2","3","^","1","+"]));             // 9

File: problems/2025/09-16/postfix-evaluation/Solution.java
import java.util.*;
public class Solution {
    public int evaluatePostfix(String[] arr) {
        Deque<Integer> st = new ArrayDeque<>();
        for (String tok : arr) {
            switch (tok) {
                case "+": {
                    int b = st.pop(), a = st.pop();
                    st.push(a + b);
                    break;
                }
                case "-": {
                    int b = st.pop(), a = st.pop();
                    st.push(a - b);
                    break;
                }
                case "*": {
                    int b = st.pop(), a = st.pop();
                    st.push(a * b);
                    break;
                }
                case "/": {
                    int b = st.pop(), a = st.pop();
                    // floor division toward -infinity
                    int q = (int)Math.floor((double)a / (double)b);
                    st.push(q);
                    break;
                }
                case "^": {
                    int b = st.pop(), a = st.pop();
                    st.push((int)Math.pow(a, b));
                    break;
                }
                default:
                    st.push(Integer.parseInt(tok));
            }
        }
        return st.peek();
    }
    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.evaluatePostfix(new String[]{"2","3","1","*","+","9","-"})); // -4
        System.out.println(s.evaluatePostfix(new String[]{"2","3","^","1","+"})); // 9
    }
}
